src/Zicht/Itertools/util/Conversions.php

```diff
@@ -23,48 +22,136 @@ class Conversions
      *
      * @param array|string|\Iterator $iterable
      * @return \Iterator
-     * @deprecated Use \Zicht\Itertools\conversions\mixed_to_iterator, will be removed in version 3.0
      */
     public static function mixedToIterator($iterable)
     {
-        return conversion_functions\mixed_to_iterator($iterable);
+        // NULL is often used to indicate that nothing is there,
+        // for robustness we will deal with NULL as it is an empty array
+        if (is_null($iterable)) {
+            $iterable = new \ArrayIterator([]);
+        }
+
+        // an array is *not* an instance of Traversable (as it is not an
+        // object and hence can not 'implement Traversable')
+        if (is_array($iterable)) {
+            $iterable = new \ArrayIterator($iterable);
+        }
+
+        // a string is considered iterable in Python
+        if (is_string($iterable)) {
+            $iterable = new StringIterator($iterable);
+        }
+
+        // a doctrine Collection (i.e. Array or Persistent) is also an iterator
+        if ($iterable instanceof Collection) {
+            $iterable = $iterable->getIterator();
+        }
+
+        if ($iterable instanceof \Traversable and !($iterable instanceof \Iterator)) {
+            $iterable = new \IteratorIterator($iterable);
+        }
+
+        // by now it should be an Iterator, otherwise throw an exception
+        if (!($iterable instanceof \Iterator)) {
+            throw new \InvalidArgumentException('Argument $ITERABLE must be a Traversable');
+        }
+
+        return $iterable;
     }
 
     /**
      * Try to transforms something into a Closure.
      *
-     * When $CLOSURE is null the returned Closure behaves like an identity function,
+     * When $closure is null the returned Closure behaves like an identity function,
      * i.e. it will return the value that it is given.
      *
      * @param null|\Closure $closure
      * @return \Closure
-     * @deprecated Use \Zicht\Itertools\conversions\mixed_to_closure, will be removed in version 3.0
      */
     public static function mixedToClosure($closure)
     {
-        return conversion_functions\mixed_to_closure($closure);
+        if (is_null($closure)) {
+            return function ($value) {
+                return $value;
+            };
+        }
+
+        if (!($closure instanceof \Closure)) {
+            // A \Closure is always callable, but a callable is not always a \Closure.
+            // Checking within this if statement is a slight optimization, preventing an unnecessary function wrap
+            if (is_callable($closure)) {
+                $closure = function () use ($closure) {
+                    return call_user_func_array($closure, func_get_args());
+                };
+            } else {
+                throw new \InvalidArgumentException('Argument $CLOSURE must be a Closure');
```

Comment:
Possibly useful to add `gettype`, or, in PHP 8, `get_debug_type` ?

Reply:
I've added the same feedback to the `mixedToIterator` call